waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/tunneling/TunnelingMetaStoreClientFactoryTest.java

```diff
@@ -0,0 +1,154 @@
+  private @Mock HiveMetaStoreClientSupplier hiveMetaStoreClientSupplier;
+  private @Mock HiveConf localHiveConf;
+  private @Mock LocalPortFactory localPortFactory;
+  private HiveConf hiveConf;
+  private TunnelingMetaStoreClientFactory tunnelingMetaStoreClientFactory;
+  private @Mock HiveMetaStoreClientSupplierFactory hiveMetaStoreClientSupplierFactory;
+
+  private final String metastoreUri = "thrift://metastore-host:42";
+  private final String name = "test";
+  private final int reconnectionRetries = 10;
+  private final String localHost = "my-machine";
+  private final int localPort = 42;
+
+  @Before
+  public void init() {
+    hiveConf = metastore.conf();
+    when(tunnelableFactorySupplier.get(any(HiveConf.class))).thenReturn(tunnelableFactory);
+    when(localHiveConfFactory.createLocalHiveConf(any(String.class), any(Integer.class), eq(hiveConf)))
+        .thenReturn(localHiveConf);
+    when(localHiveConf.getVar(HiveConf.ConfVars.METASTOREURIS)).thenReturn(metastoreUri);
+    when(hiveMetaStoreClientSupplierFactory.newInstance(localHiveConf, name, reconnectionRetries))
+        .thenReturn(hiveMetaStoreClientSupplier);
+    when(localPortFactory.getLocalPort()).thenReturn(localPort);
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_PRIVATE_KEYS.varname, "private_key");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_KNOWN_HOSTS.varname, "");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_STRICT_HOST_KEY_CHECKING.varname, "yes");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_ROUTE.varname, "user@hop1 -> hop2");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_LOCALHOST.varname, localHost);
+
+    tunnelingMetaStoreClientFactory = new TunnelingMetaStoreClientFactory(tunnelableFactorySupplier,
+        metaStoreClientFactory, localHiveConfFactory, hiveMetaStoreClientSupplierFactory, localPortFactory);
+
+  }
+
+  @Test
+  public void newInstanceWithTunneling() throws Exception {
+    ArgumentCaptor<TunnelableSupplier> captor = ArgumentCaptor.forClass(TunnelableSupplier.class);
+    tunnelingMetaStoreClientFactory.newInstance(hiveConf, name, reconnectionRetries);
+    verify(tunnelableFactory)
+        .wrap(captor.capture(), eq(tunnelingMetaStoreClientFactory.METHOD_CHECKER), eq("my-machine"), anyInt(),
+            eq("localhost"), eq(metastore.getThriftPort()));
+  }
+
+  @Test
+  public void localHiveConfigUsesCorrectParameters() {
+    tunnelingMetaStoreClientFactory.newInstance(hiveConf, name, reconnectionRetries);
+    ArgumentCaptor<String> stringArgument = ArgumentCaptor.forClass(String.class);
+    ArgumentCaptor<Integer> intArgument = ArgumentCaptor.forClass(Integer.class);
+    verify(localHiveConfFactory).createLocalHiveConf(stringArgument.capture(), intArgument.capture(), eq(hiveConf));
+    assertThat(stringArgument.getValue(), is(localHost));
+    assertThat(intArgument.getValue(), is(localPort));
+  }
+
+  @Test
+  public void tunnelableFactorySupplierUsesCorrectParameters() {
```

Comment:
These tests are too fragmented, you making a test to verify that a line is called that doesn't add much imo. Look at the number of conditionals in the code thats the number of tests you should have to get full coverage.
Got one if, means you need two tests as there are two possible paths code can take and you want to run both in your test.

Reply:
Some of these tests are to prevent previous bugs from happening again. @courtsVII told me that at one point in the past, the code was changed and bugs were introduced because `hiveConf` was used instead of `localConf`. So while less tests would be enough for this class, I thought that being more defensive was the right approach.

Reply:
sure there are always exception but I think you went a bit too far, have a look at my other comment and if you rewrite some of the tests following those suggestions see if these small ones are actually adding anything or needed.